protocol/messenger_mention.go

```
@@ -0,0 +1,1115 @@
+package protocol
+
+// this is a reimplementation of the mention feature in status-react
+// reference implementation: https://github.com/status-im/status-react/blob/972347963498fc4a2bb8f85541e79ed0541698da/src/status_im/chat/models/mentions.cljs#L1
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/status-im/status-go/api/multiformat"
+	"github.com/status-im/status-go/protocol/common"
+)
+
+const (
+	endingChars = `[\s\.,;:]`
+
+	charAtSign     = "@"
+	charQuote      = ">"
+	charNewline    = "\n"
+	charAsterisk   = "*"
+	charUnderscore = "_"
+	charTilde      = "~"
+	charCodeBlock  = "`"
+
+	intUnknown = -1
+)
+
+var (
+	specialCharsRegex = regexp.MustCompile("[@~\\\\*_\n>`]{1}")
+	endingCharsRegex  = regexp.MustCompile(endingChars)
+	wordRegex         = regexp.MustCompile("^[\\w\\d]*" + endingChars + "|^[\\S]*$")
+)
+
+type specialCharLocation struct {
+	Index int
+	Value string
+}
+
+type atSignIndex struct {
+	Pending []int
+	Checked []int
+}
+
+type styleTag struct {
+	Len int
+	Idx int
+}
+
+type textMeta struct {
+	atSign         *atSignIndex
+	styleTagMap    map[string]*styleTag
+	quoteIndex     *int
+	newlineIndexes []int
+}
+
+type MentionableUser struct {
+	*Contact
+
+	primaryName       string
+	secondaryName     string
+	searchablePhrases []string
+
+	Key          string // used by :key-fn in react
+	Match        string
+	SearchedText string
+}
+
+func (c *MentionableUser) MarshalJSON() ([]byte, error) {
+	contactJSONMap := make(map[string]interface{})
+	contactJSONMap["primaryName"] = c.PrimaryName()
+	contactJSONMap["secondaryName"] = c.SecondaryName()
+	contactJSONMap["ensVerified"] = c.ENSVerified
+	contactJSONMap["added"] = c.added()
+	contactJSONMap["displayName"] = c.GetDisplayName()
+	contactJSONMap["key"] = c.Key
+	contactJSONMap["match"] = c.Match
+	contactJSONMap["searchedText"] = c.SearchedText
+	contactJSONMap["id"] = c.ID
+	compressedKey, err := multiformat.SerializeLegacyKey(c.ID)
```

I think the code could be improved a little bit if we defined a named struct or if we used an inline struct (with/without JSON tags). We don't need to give up on type safety by using `interface{}`. We can also move the compressed key line to the top, so we can fail fast.

---

fixed